pkg/driver/nydus/nydus.go

```diff
@@ -433,150 +436,299 @@ func (d *Driver) getChunkDict(ctx context.Context, provider accelcontent.Provide
+func (d *Driver) PushRemoteCache(ctx context.Context, pvd accelcontent.Provider, ref string, cacheIndex *ocispec.Index) error {
+	for _, manifest := range cacheIndex.Manifests {
+		if err := pvd.Push(ctx, manifest, ref); err != nil {
+			return err
+		}
+	}
+	manifestIndexDesc, manifestIndexBytes, err := nydusutils.MarshalToDesc(*cacheIndex, ocispec.MediaTypeImageIndex)
+	if err != nil {
+		return errors.Wrap(err, "marshal remote cache manifest index")
+	}
+	if err = content.WriteBlob(ctx, pvd.ContentStore(), ref, bytes.NewReader(manifestIndexBytes), *manifestIndexDesc); err != nil {
+		return errors.Wrap(err, "write remote cache manifest index")
+	}
+	if err = pvd.Push(ctx, *manifestIndexDesc, ref); err != nil {
+		return err
 	}
 	return nil
 }
 
-// PushRemoteCache update cache manifest and push to remote
-func (d *Driver) PushRemoteCache(ctx context.Context, pvd accelcontent.Provider, ref string) error {
+// UpdateRemoteCache update cache layer from upper to lower
+func (d *Driver) UpdateRemoteCache(ctx context.Context, provider accelcontent.Provider, ref string, orgDesc, newDesc, cacheDesc *ocispec.Descriptor) (*ocispec.Index, error) {
+	cs := provider.ContentStore()
+	cacheLayers := map[*platforms.Platform][]ocispec.Descriptor{}
+
+	switch orgDesc.MediaType {
+	case ocispec.MediaTypeImageManifest, images.MediaTypeDockerSchema2Manifest:
+		targetLayers, err := getNydusCacheLayers(ctx, cs, *orgDesc, *newDesc)
+		if err != nil {
+			return nil, err
+		}
+		// platform of original or new image maybe lost, get from config platform
+		platform, err := images.Platforms(ctx, cs, *orgDesc)
+		if err != nil {
+			return nil, err
+		}
+		cacheLayers[&platform[0]] = targetLayers
+
+	case images.MediaTypeDockerSchema2ManifestList, ocispec.MediaTypeImageIndex:
+		orgManifests, err := utils.GetManifests(ctx, cs, *orgDesc, d.platformMC)
+		if err != nil {
+			return nil, errors.Wrap(err, "get original manifest list")
+		}
+		newManifests, err := utils.GetManifests(ctx, cs, *newDesc, d.platformMC)
+		if err != nil {
+			return nil, errors.Wrap(err, "get new manifest list")
+		}
+		for _, newManifestDesc := range newManifests {
+			targetLayers := []ocispec.Descriptor{}
+			if newManifestDesc.Platform == nil {
+				newManiPlatforms, err := images.Platforms(ctx, cs, newManifestDesc)
+				if err != nil {
+					return nil, errors.Wrap(err, "get converted manifest platforms")
+				}
+				newManifestDesc.Platform = &newManiPlatforms[0]
+			}
+			// find original manifest matches converted manifest's platform
+			matcher := platforms.NewMatcher(*newManifestDesc.Platform)
+			for _, orgManifestDesc := range orgManifests {
+				if orgManifestDesc.Platform == nil {
+					orgManiPlatforms, err := images.Platforms(ctx, cs, orgManifestDesc)
+					if err != nil {
+						return nil, errors.Wrap(err, "get original manifest platforms")
+					}
+					orgManifestDesc.Platform = &orgManiPlatforms[0]
+				}
+				if matcher.Match(*orgManifestDesc.Platform) {
```

Comment:
Why not use `&orgManiPlatforms[0]` directly here?

Reply:
ok